### PR TITLE
feat(kubectl) add fish shell completion

### LINKED ISSUE #30763: Add support for fish shell autocompletion
**FEATURE REQUEST**:

[Fish shell](https://fishshell.com) is a relatively new modern interactive user friendly command line shell. It really shines at user friendly features like visual tab completion menus, live syntax highlighting and other useful features.

`kubectl` has a massive set of sub-commands and options. Remembering all those is a chore. Specially when getting to know Kubernetes. All the help is appreciated.

Please provide autocompletion support for fish shell in addition to bash and zsh.

Thanks

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.4+dd6b458", GitCommit:"dd6b458ef8dbf24aff55795baa68f83383c9b3a9", GitTreeState:"not a git tree", BuildDate:"2016-08-04T09:44:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"darwin/amd64"}
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

**Environment**:
**Others**: Fish shell

### PR DESCRIPTION
**Pull request by mmr**
Adds `fish` shell completion to `kubectl`
Fixes #30763

`cobra` -the lib used to generate `bash` and `zsh` completions- do not support `fish` (yet)
That is why the `fish` completion script is a lot longer and more complex

All the completion code was created by https://github.com/evanlucas/fish-kubectl-completions
Thank you @evanlucas

```release-note
Add fish shell completion to kubectl
```

### CODE REVIEW

File: pkg/kubectl/cmd/completion.go [hunk 1]
```diff
@@ -308,3 +314,1203 @@ _complete kubectl 2>/dev/null
...
+    for i in (__kubectl_pods --no-prefix pods)
+        if string match -q "*$i*" -- $cmd
+            return 1
+        end
+    end
+    return 0
+end
+
+function __kubectl_resource_types
+    echo clusters
```
Comment (sttts) [inline pkg/kubectl/cmd/completion.go]: can't these be generated instead of a hard-coded list?

File: pkg/kubectl/cmd/completion.go [hunk 2]
```diff
@@ -308,3 +314,1203 @@ _complete kubectl 2>/dev/null
...
+            echo serviceaccount\t"Create a service account with the specified name"
+        case set
+            echo image\t"Update image of a pod template"
+            echo resources\t"Update resource requests/limits on objects with pod templates"
+    end
+end
+
+# Global command-line options:
+
+complete -c kubectl -l "alsologtostderr" -d "Log to standard error as well as files"
```
Comment (sttts) [inline pkg/kubectl/cmd/completion.go]: can't these be autogenerated?

**Comment by sttts** [comment]
> All the completion code was created by https://github.com/evanlucas/fish-kubectl-completions

what does this mean? If there is auto-generation already, can't this be done at runtime of the `kubectl completion` command?

In general, I fear that this code is hard to maintain. We need more automation. A new command or new flag has to appear automatically in the completion.

**Comment by cblecker** [comment]
Yeah, I don't think it's sustainable to maintain fish completion without it being auto-generated in some way. Looking at that repo, it doesn't look like it's auto-generated either.

File: pkg/kubectl/cmd/completion.go (hunk 1 above)
Comment (iax7) [reply pkg/kubectl/cmd/completion.go]: yes, with this code:
```
kubectl get 2>&1 | egrep -o '\* [^ ]+' | cut -d' '  -f2
```

**Pull request closed without merging**